Fix LSP References Single-Result Jump Behavior

jls83 opened the pull request:
# Description

When there is only a single result for a call to `lsp_references`, the previous cursor position is not added to the jumplist. This is inconsistent with the behavior of other Telescope built-ins.

This PR chanes the implementation of the `lsp.references` method to use the `list_or_jump` helper function already in use for other LSP-related items (e.g. `definitions`, `type_definitions`, `implementations`). It also adds an optional results filter to `list_or_jump` via the `results_filter` attribute in `opts`.

## Type of change
- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list relevant details about your configuration

- [ ] Find a symbol that has a single usage & call `require('telescope.builtin').lsp_references()`. Try jumping back (via `<C-o>`) to the previous cursor position.
- [ ] Confirm symbols with multiple usages work as intended.

**Configuration**:
* Neovim version (nvim --version): v0.8.1
* Operating system and version: Debian (testing)

# Checklist:

- [x] My code follows the style guidelines of this project (stylua)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (lua annotations)

jls83 commented:
@Conni2461 sorry for the ping, but do I need to directly assign someone for the CI workflow and/or a review? Thanks!

Conni2461 commented on lua/telescope/builtin/__lsp.lua:
```diff
@@ -179,6 +114,10 @@ local function list_or_jump(action, title, opts)
       vim.list_extend(flattened_results, result)
     end
 
+    if not vim.F.if_nil(opts.results_filter) then
```
we should check if `opts.results_filter` is a function with `type`.

jls83 replied:
Good call, done.

Conni2461 commented on lua/telescope/builtin/__lsp.lua:
```diff
@@ -211,6 +150,22 @@ local function list_or_jump(action, title, opts)
+    end
+  end
+
+  return list_or_jump("textDocument/references", "LSP References", opts)
```
```lua
params.context = { includeDeclaration = vim.F.if_nil(opts.include_declaration, true) }
```

we can either pass in params as opts.

jls83 replied:
Not sure what this is referring to, sorry!

gbroques replied:
I think this comment is referring to the need to support the `include_declaration` option for `lsp_references`.

See the following code:
https://github.com/nvim-telescope/telescope.nvim/blob/5fff2a138b04be7e23959c1da5e535bdac7e1518/lua/telescope/builtin/__lsp.lua#L17-L19

jls83 replied:
@gbroques thanks for the clarification! I added some logic in `list_or_jump` to handle this case.

Conni2461 commented on lua/telescope/builtin/__lsp.lua:
```diff
@@ -9,71 +9,6 @@ local utils = require "telescope.utils"
-      elseif opts.jump_type == "vsplit" then
-        vim.cmd "vnew"
-      end
-      -- jump to location
-      local location = locations[1]
-      local bufnr = opts.bufnr
-      if location.filename then
-        bufnr = vim.uri_to_bufnr(vim.uri_from_fname(location.filename))
-      end
-      vim.api.nvim_win_set_buf(0, bufnr)
-      vim.api.nvim_win_set_cursor(0, { location.lnum, location.col - 1 })
```
maybe look at: https://github.com/nvim-telescope/telescope.nvim/pull/2229#issuecomment-1311109346

I've not check what they said but if you say it works, i'll will do some investigating/testing

Conni2461 replied:
Maybe something changed upstream and now the objects are the same. But we still have 0.7 support so we need to make sure it works for both versions.

jls83 replied:
I didn't run into the problem in the linked issue, but I've been making changes with NeoVim v0.8.1 so you might be right.

jls83 replied:
@Conni2461 let me know if anything needs to be updated here.

Conni2461 commented:
I am sorry about the late review. There is no excuse i should have done it a long time ago :|

Thanks for the PR :) I've also thought that using the same function should be possible, so thanks for implementing it.

I've done a quick review. `opts.results_filter` kinda gives us a new feature that should be documented for the all pickers that use `list_or_jump`. Do you know how to do that?

jls83 commented:
No worries on the delay! End of the year is always a busy time :)

I added docs for the affected pickers, but I'm not sure if anything else needs to be done besides the CI pushing the changes to the branch once the workflow permissions are granted. Also, I noticed a subtle bug with my initial implementation -- if someone passed in a `results_filter` fn for the `lsp_references` picker, it would be discarded in favor of the "remove current line" filter. I added some extra logic to run both filters if available.

Let me know if you need anything else!

gbroques commented:
Thank you for the work on this @Conni2461 and @jls83.

Any chance we could move this PR along?

If not, I'd be willing to submit a simpler PR that only changes the implementation of the `lsp.references` method to use the `list_or_jump` helper function already in use for other LSP-related items (e.g. definitions, type_definitions, implementations).

We could maybe add the `results_filter` if we want it later.

nieomylnieja commented:
Upping this one :point_up: could also help if need be :)

Conni2461 commented:
uff thats a 2022 PR, i'll finish it this weekend. thanks for the ping

jls83 commented:
@Conni2461 I fixed up my branch with master & made some of the requested changes. Let me know if you need anything else!
